src/lib/hardware/pci.c

```diff
@@ -54,9 +54,14 @@ uint32_t volatile *map_pci_resource(int fd)
   }
 }
 
-void close_pci_resource(int fd)
+void close_pci_resource(int fd, uint32_t volatile *addr)
```

Comment:
About the `volatile` keword, it's only there to be consistent with the return value of `map_pci_resource()`.  I agree that it's superfluous and would be better removed.  Even in C it's dubious that would be any benefit as an argument or return value.

About barriers in initialization code, most of the places where order is important, the Intel docs advise to insert a wait, sometimes a few microseconds, sometimes up to a millisecond.  In those cases we use `C.usleep()`, which should have the barrier effect as well.  (and it's hard to imagine some reordering/pending write to last as long as a microsecond).

Where we could be missing some defensive coding is more in the data handling itself.  Maybe the `M_sf:sync_transmit ()` race is related to this?

Finally, I don't think the initialization code is ever compiled.  Not even in the `manyreconf()` calls it tightly enough.  What would be a simple way to check that?